### PR TITLE
dnsdisc/client.nim: use '{.async.}' in callback Resolver definition

### PR DESCRIPTION
**Pull request by Ivansete-status**

### CODE REVIEW

File: tests/test_client.nim [hunk 1]
```diff
@@ -11,16 +11,23 @@ procSuite "Test DNS Discovery: Client":
...
   treeRecords["C7HRFPF3BLGF3YR4DY5KX3SMBE.nodes.example.org"] = "enrtree://AM5FCQLWIZX2QFPNJAP7VUERCCRNGRHWZG3YYHIUV7BVDQ5FDPRT2@morenodes.example.org"
   treeRecords["JWXYDBPXYWG6FX3GMDIBFA6CJ4.nodes.example.org"] = "enrtree-branch:2XS2367YHAXJFGLZHVAWLQD4ZY,H4FHT4B454P6UXFD7JCYQ5PWDY,MHTDO6TMUBRIA2XWG5LUDACK24"
   treeRecords["2XS2367YHAXJFGLZHVAWLQD4ZY.nodes.example.org"] = "enr:-HW4QOFzoVLaFJnNhbgMoDXPnOvcdVuj7pDpqRvh6BRDO68aVi5ZcjB3vzQRZH2IcLBGHzo8uUN3snqmgTiE56CH3AMBgmlkgnY0iXNlY3AyNTZrMaECC2_24YYkYHEgdzxlSNKQEnHhuNAbNlMlWJxrJxbAFvA"
   treeRecords["H4FHT4B454P6UXFD7JCYQ5PWDY.nodes.example.org"] = "enr:-HW4QAggRauloj2SDLtIHN1XBkvhFZ1vtf1raYQp9TBW2RD5EEawDzbtSmlXUfnaHcvwOizhVYLtr7e6vw7NAf6mTuoCgmlkgnY0iXNlY3AyNTZrMaECjrXI8TLNXU0f8cthpAMxEshUyQlK-AM0PW2wfrnacNI"
   treeRecords["MHTDO6TMUBRIA2XWG5LUDACK24.nodes.example.org"] = "enr:-HW4QLAYqmrwllBEnzWWs7I5Ev2IAs7x_dZlbYdRdMUx5EyKHDXp7AV5CkuPGUPdvbv1_Ms1CPfhcGCvSElSosZmyoqAgmlkgnY0iXNlY3AyNTZrMaECriawHKWdDRk2xeZkrOXBQ0dfMFLHY4eENZwdufn1S1o"
 
-  proc resolver(domain: string): Future[string] {.async.} =
-    return treeRecords[domain]
+  proc resolver(domain: string): Future[string] {.gcsafe, raises: [].} =
```
Comment (arnetheduck) [inline tests/test_client.nim]: are these changes needed? they should not be

File: tests/test_client.nim [hunk 2]
```diff
@@ -101,8 +108,10 @@ procSuite "Test DNS Discovery: Client":
       links.contains(parseLinkEntry("enrtree://AM5FCQLWIZX2QFPNJAP7VUERCCRNGRHWZG3YYHIUV7BVDQ5FDPRT2@morenodes.example.org").tryGet())
     
     # Invalid case
-    proc invalidResolver(domain: string): Future[string] {.async.} =
-      return ""
+    proc invalidResolver(domain: string): Future[string] {.gcsafe, raises: [].} =
```
Comment (arnetheduck) [inline tests/test_client.nim]: ditto

File: discovery/dnsdisc/client.nim [hunk 3]
```diff
@@ -29,7 +29,7 @@ type
   
   ## A Resolver proc takes a DNS domain as argument and
   ## returns the TXT record at that domain
-  Resolver* = proc(domain: string): Future[string] {.gcsafe.}
+  Resolver* = proc(domain: string): Future[string] {.gcsafe, raises: [].}
```
Comment (arnetheduck) [inline discovery/dnsdisc/client.nim]: :+1: 

callbacks that return future must always be `{.async.}` or equivalently `{.gcsafe, raises: [].}` - the former is preferred but the latter also works

**Comment by Ivansete-status** [comment]
Thanks for the comments @arnetheduck ! I think it is better now :D

**Review by arnetheduck — APPROVED** [review]
